File: web/node/scripts/flutter_wallet_connect.js
const { createWeb3Modal, defaultConfig } = require('@web3modal/ethers');
const { BrowserProvider, Contract, formatUnits } = require('ethers');
const ICO = require('./ico.js');
const ERC20 = require('./erc20.js');

FlutterWalletConnect = function ({ projectId, metadata, chains }) {
    const config = defaultConfig({
        metadata,
        enableInjected: false
    });

    this.web3Modal = createWeb3Modal({
        projectId: projectId,
        ethersConfig: config,
        chains: chains,
    });

    this.connect = async function () {
        return this.web3Modal.open('Connect');
    }

    this.disconnect = function () {
        return this.web3Modal.disconnect();
    }

    this.getAccounts = async function () {
        return this.web3Modal.getAddress();
    }

    this.getChainId = function () {
        return this.web3Modal.getChainId();
    }

    this.isConnected = function () {
        return this.web3Modal.getIsConnected();
    }

    this.subscribeWalletInfo = function (callback) {
        this.web3Modal.subscribeWalletInfo(callback);
    }

    this.subscribeProvider = function (callback) {
        this.web3Modal.subscribeProvider(callback);
    }

    this.subscribeState = function (callback) {
        this.web3Modal.subscribeState(callback);
    }

    this.switchNetwork = async function (chainId) {
        return this.web3Modal.switchNetwork(chainId);
    }

    this.closeModal = async function () {
        return this.web3Modal.close();
    }

    this.openModal = async function (view) {
        return this.web3Modal.open({ view: view });
    }

    this.getState = function () {
        return this.web3Modal.getState();
    }

    this.getProvider = async function () {
        return this.web3Modal.getWalletProvider();
    }

    this.icoSmartContract = async function () {
        const provider = await this.getProvider();
        const ico = new ICO({
            walletConnectProvider: provider,
            contractAddress: '0x80e2dDD5fB4acB62755e1eD645bB8819029b0766',
        });
        await ico.init();
        return ico;
    }

    this.usdtSmartContract = async function () {
        const provider = await this.getProvider();
        const usdt = new ERC20({
            walletConnectProvider: provider,
            contractAddress: '0xc4a0879BBFbC7f557De9cfB4ACB2Dbf6441188BC',
        });
        await usdt.init();
        return usdt;
    }

    this.ppcbSmartContract = async function () {
        const provider = await this.getProvider();
        const ppcb = new ERC20({
            walletConnectProvider: provider,
            contractAddress: '0xbe9469fA5Ad0e37aD74Be8F71e3C661e94E61882',
        });
        await ppcb.init();
        return ppcb;
    }
}

window.FlutterWalletConnect = FlutterWalletConnect;

File: web/lib/flutter_wallet_connect/flutter_wallet_connect.dart
// ignore_for_file: library_private_types_in_public_api

@JS('window')
import 'package:js/js.dart';
import 'package:js/js_util.dart';

part 'wallet_connect_metadata.dart';
part 'network_chain.dart';
part 'connect_wallet_info.dart';
part 'ether_store_state.dart';
part 'provider.dart';
part 'web3modal_state.dart';
part 'erc20.dart';
part 'ico.dart';

@JS('FlutterWalletConnect')
class _FlutterWalletConnect {
  external _FlutterWalletConnect(FlutterWalletConnectOptions options);
  external dynamic get web3Modal;
  external Future<void> connect();
  external Future<void> disconnect();
  external int getChainId();
  external bool isConnected();
  external Future<String?> getAccounts();
  external Future<void> switchNetwork(int chainId);
  external Future<void> closeModal();

  // view: 'Account' | 'Connect' | 'Networks' | 'ApproveTransaction' | 'OnRampProviders';
  external Future<void> openModal(String view);
  external Web3ModalState getState();
  external void subscribeWalletInfo(void Function(ConnectWalletInfo?) callback);
  external void subscribeProvider(void Function(EthersStoreState?) callback);
  external void subscribeState(void Function(Web3ModalState?) callback);
  external Future<dynamic> getProvider();
  external Future<ICO> icoSmartContract();
  external Future<ERC20> usdtSmartContract();
  external Future<ERC20> ppcbSmartContract();
}

class FlutterWalletConnect implements _FlutterWalletConnect {
  FlutterWalletConnect(FlutterWalletConnectOptions options)
      : instance = _FlutterWalletConnect(options);
  final _FlutterWalletConnect instance;

  @override
  Future<String?> getAccounts() {
    return promiseToFuture<String?>(instance.getAccounts());
  }

  @override
  void subscribeWalletInfo(void Function(ConnectWalletInfo?) callback) {
    instance.subscribeWalletInfo(allowInterop(callback));
  }

  @override
  void subscribeProvider(void Function(EthersStoreState?) callback) {
    instance.subscribeProvider(allowInterop(callback));
  }

  @override
  void subscribeState(void Function(Web3ModalState?) callback) {
    instance.subscribeState(allowInterop(callback));
  }

  @override
  dynamic get web3Modal => instance.web3Modal;

  @override
  Future<void> connect() {
    return promiseToFuture<void>(instance.connect());
  }

  @override
  Future<void> disconnect() {
    return promiseToFuture<void>(instance.disconnect());
  }

  @override
  bool isConnected() {
    return instance.isConnected();
  }

  @override
  int getChainId() {
    return instance.getChainId();
  }

  @override
  Future<void> switchNetwork(int chainId) {
    return promiseToFuture<void>(instance.switchNetwork(chainId));
  }

  @override
  Future<void> closeModal() {
    return promiseToFuture<void>(instance.closeModal());
  }

  // view: 'Account' | 'Connect' | 'Networks' | 'ApproveTransaction' | 'OnRampProviders';
  @override
  Future<void> openModal(String view) {
    return promiseToFuture<void>(instance.openModal(view));
  }

  @override
  Web3ModalState getState() {
    return instance.getState();
  }

  @override
  Future<dynamic> getProvider() {
    return promiseToFuture<dynamic>(instance.getProvider());
  }

  @override
  Future<ICO> icoSmartContract() async {
    final ICO ico = await promiseToFuture<ICO>(instance.icoSmartContract());
    return ICOImp(ico);
  }

  @override
  Future<ERC20> ppcbSmartContract() async {
    final ERC20 ppcb = await promiseToFuture<ERC20>(instance.ppcbSmartContract());
    return ERC20Imp(ppcb);
  }

  @override
  Future<ERC20> usdtSmartContract() async {
    final ERC20 usdt = await promiseToFuture(instance.usdtSmartContract());
    return ERC20Imp(usdt);
  }
}

@JS()
@anonymous
class FlutterWalletConnectOptions {
  external factory FlutterWalletConnectOptions({
    required String projectId,
    required WalletConnectMetadata metadata,
    required List<NetworkChain> chains,
  });

  external String get projectId;
  external WalletConnectMetadata get metadata;
  external List<NetworkChain> get chains;
}
